iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp

```diff
@@ -0,0 +1,155 @@
+///
+///         // Usage with l-values
+///         // Pitfall: Ensure that lifetime of callable suits the point in time of calling callback()
+///         auto callable = [&]{ doSomething(); };
+///         cxx::function_ref<void()> callback(callable);
+///         // Call the callback
+///         callback();
+/// @endcode
+
+template <typename SignatureType>
+class function_ref;
+
+template <class ReturnType, class... ArgTypes>
+class function_ref<ReturnType(ArgTypes...)>
+{
+    using SignatureType = ReturnType(ArgTypes...);
+    using SelfType = function_ref<SignatureType>;
+    template <typename T>
+    using EnableIfNotFunctionRef = typename std::enable_if<!std::is_same<std::decay<T>, function_ref>::value>::type;
+
+  public:
+    /// @brief Creates an empty function_ref
+    function_ref() noexcept
+        : m_target(nullptr)
+        , m_functionPointer(nullptr)
+    {
+    }
+
+    /// @brief Creates an empty function_ref
+    function_ref(nullptr_t) noexcept
+        : function_ref()
+    {
+    }
+
+    /// @brief D'tor
+    ~function_ref() noexcept = default;
+
+    /// @brief Copy c'tor
+    function_ref(const function_ref&) noexcept = default;
+
+    /// @brief Copy assignment operator
+    function_ref& operator=(const function_ref&) noexcept = default;
+
+    /// @brief Create a function_ref
+    template <typename CallableType, typename = EnableIfNotFunctionRef<SelfType>>
+    function_ref(CallableType&& callable) noexcept
+        : m_target(reinterpret_cast<void*>(std::addressof(callable)))
+        , m_functionPointer([](void* target, ArgTypes... args) -> ReturnType {
+            return (*reinterpret_cast<typename std::add_pointer<CallableType>::type>(target))(
+                std::forward<ArgTypes>(args)...);
+        })
+    {
+    }
+
+    /// @brief Moves a function_ref
+    function_ref(function_ref&& rhs) noexcept
+    {
+        *this = std::move(rhs);
+    }
+
+    /// @brief Move assignment operator
+    function_ref& operator=(function_ref&& rhs) noexcept
+    {
+        if (this != &rhs)
+        {
+            m_target = rhs.m_target;
+            m_functionPointer = rhs.m_functionPointer;
+            // Make sure no UB can happen by marking the lvalue as invalid
+            rhs.m_target = nullptr;
+            rhs.m_functionPointer = nullptr;
+        }
+        return *this;
+    };
+
+    /// @brief Calls the provided callable
+    ReturnType operator()(ArgTypes... args) const noexcept
+    {
+        if (!m_target)
+        {
+            // Callable was called without user having assigned one beforehand
+            std::terminate();
+        }
+        return m_functionPointer(m_target, std::forward<ArgTypes>(args)...);
+    }
+
+    /// @brief Checks whether a valid target is contained
+    explicit operator bool() const noexcept
+    {
+        return m_target != nullptr;
+    }
+
+    void swap(function_ref& rhs) noexcept
+    {
+        std::swap(m_target, rhs.m_target);
```

Comment:
is std::swap part of the ASILD subset of the STL we can use?

Reply:
I dont think so but it should be feasible to implement this on our own